File: peerrpc-go/transport/channel.go
// Package transport wraps a WebRTC DataChannel into a reliable
// length-prefixed Frame reader/writer with built-in sharding for large
// messages and outbound backpressure.
//
// Layering (top to bottom):
//
//	rpc.Server / rpc.Client
//	      │  Frame / ResponseFrame (proto messages)
//	      ▼
//	transport.Channel.Send / Recv
//	      │  transparent Chunk split / reassembly
//	      ▼
//	peer.DataChannel (ordered, reliable)
//
// The transport exposes two channels:
//   - SendFrame(frame proto.Message) blocks under backpressure.
//   - RecvFrame() returns decoded Frame / ResponseFrame.
package transport

import (
	"context"
	"errors"
	"fmt"
	"io"
	"sync"
	"time"

	"github.com/pion/webrtc/v4"
	"google.golang.org/protobuf/proto"
)

// SCTP / framing thresholds.
//
// WebRTC DataChannels negotiate an SCTP max-message-size (256 KiB by
// default, per pion's a=max-message-size:262144 and the browser). Each
// dc.send() carries one ENCODED frame = 4-byte length prefix + protobuf
// envelope (Routing + Data tags) + payload. That envelope is ~40 bytes
// for a Chunk frame, so a payload of exactly 256 KiB would push the
// encoded frame just over the limit and the browser rejects it with
// "Trying to send message larger than max-message-size".
//
// We reserve frameOverhead so the encoded frame stays safely under the
// negotiated max. These are transport-layer chunking thresholds only;
// they are NOT carried on the wire (Chunk carries total_size/offset/
// data), so each side may pick its own value without a protocol change.
const (
	// maxFrameBytes is the negotiated SCTP max-message-size. Every
	// encoded frame MUST stay at or below this. peerrpc peers advertise
	// a=max-message-size:262144 in their local SDPs (pion already
	// defaults to 262144; webrtc-rs needs the SDP injection plus
	// can_send=Unbounded on its SettingEngine) so the negotiation
	// reaches 256 KiB. Once both sides advertise ≥256 KiB, chunk sizes
	// of 255 KiB are safe.
	maxFrameBytes = 256 * 1024
	// frameOverhead is the worst-case bytes added by the 4-byte length
	// prefix plus the protobuf Frame/Routing/Data/Chunk envelope. 1 KiB
	// is far larger than the real ~40 bytes, giving a safety margin.
	frameOverhead = 1 * 1024

	// InlineMax is the largest payload that may ride in Call.inline_data
	// or Begin.inline_data. Above this, frames use Data.message or
	// Data.chunk.
	InlineMax = 16 * 1024
	// MessageMax is the largest single Data.message payload. Larger
	// logical payloads MUST be split into Chunk frames. Sized so the
	// encoded frame fits under the SCTP max-message-size.
	MessageMax = maxFrameBytes - frameOverhead
	// ChunkSize is the per-chunk payload size when fragmenting. Sized so
	// the encoded frame fits under the SCTP max-message-size.
	ChunkSize = maxFrameBytes - frameOverhead
	// BufferedAmountHigh is the high-watermark (bytes) above which the
	// channel pauses new Send calls until the DataChannel drains.
	BufferedAmountHigh = 1 << 20 // 1 MiB
)

// DataChannelLabel carries the protocol version on the wire.
const DataChannelLabel = "peerrpc-v1"

// Channel is the transport-level duplex pipe backed by a WebRTC
// DataChannel.
//
// A Channel is goroutine-safe for Send. It is intended for a single
// reader (the multiplexer) but that reader may dispatch the decoded
// frames to many streams.
type Channel struct {
	dc *webrtc.DataChannel

	// outbound backpressure
	bufferedLow       chan struct{}
	bufferedLowActive bool
	bufferedLowMu     sync.Mutex

	// inbound queue
	recv chan []byte

	// chunk reassembly
	reasm *Reassembler

	closeOnce sync.Once
	closeMu   sync.Mutex
	closed    chan struct{}
	closeErr  error
	shutdownDone bool
}

// assembler collects chunks for one logical message on one stream.
type assembler struct {
	buf      []byte
	got      int
	total    int
	lastSeen time.Time
}

// New wraps an established DataChannel. Caller is still responsible
// for the underlying PeerConnection; this constructor only registers
// the OnMessage / OnClose handlers.
//
// The DataChannel MUST be created with ordered=true, reliable delivery
// (maxRetransmits = infinity). New does not verify this because pion's
// API makes it awkward at runtime; the peer package enforces it at
// DataChannel creation time.
func New(dc *webrtc.DataChannel) *Channel {
	c := &Channel{
		dc:          dc,
		bufferedLow: make(chan struct{}, 1),
		recv:        make(chan []byte, 256),
		reasm:       NewReassembler(),
		closed:      make(chan struct{}),
	}

	dc.OnBufferedAmountLow(func() {
		c.bufferedLowMu.Lock()
		c.bufferedLowActive = false
		c.bufferedLowMu.Unlock()
		select {
		case c.bufferedLow <- struct{}{}:
		default:
		}
	})
	dc.OnMessage(func(msg webrtc.DataChannelMessage) {
		select {
		case c.recv <- msg.Data:
		case <-c.closed:
		}
	})
	dc.OnClose(func() {
		c.shutdown(io.ErrClosedPipe)
	})

	// Configure pion to emit OnBufferedAmountLow whenever the buffered
	// amount drops below BufferedAmountThreshold. The threshold is set
	// on the DataChannel; we let the peer layer set it (or default).
	dc.SetBufferedAmountLowThreshold(BufferedAmountHigh)

	return c
}

// SendFrame marshals frame and writes it through the DataChannel
// with a length prefix matching the wire protocol. The frame format is:
//
//	uint32 BE length | protobuf payload
//
// This MUST match the TS transport's encoding so both sides can
// interoperate.
func (c *Channel) SendFrame(ctx context.Context, frame proto.Message) error {
	payload, err := proto.Marshal(frame)
	if err != nil {
		return fmt.Errorf("transport: marshal: %w", err)
	}
	return c.SendRaw(ctx, lengthPrefix(payload))
}

// lengthPrefix prepends a 4-byte big-endian length to payload.
func lengthPrefix(payload []byte) []byte {
	out := make([]byte, 4+len(payload))
	out[0] = byte(len(payload) >> 24)
	out[1] = byte(len(payload) >> 16)
	out[2] = byte(len(payload) >> 8)
	out[3] = byte(len(payload))
	copy(out[4:], payload)
	return out
}

// SendRaw writes pre-marshaled bytes verbatim through the DataChannel.
// It is the relay's primitive: the relay forwards frames without
// inspecting them, so marshaling again would be wasted CPU and would
// risk non-determinism when the wire payload was produced by another
// peer's encoder.
//
// SendRaw applies the same backpressure as SendFrame.
func (c *Channel) SendRaw(ctx context.Context, payload []byte) error {
	select {
	case <-c.closed:
		return fmt.Errorf("transport: channel closed: %w", c.closeErr)
	default:
	}

	if err := c.awaitBufferLow(ctx); err != nil {
		return err
	}
	if err := c.dc.Send(payload); err != nil {
		return fmt.Errorf("transport: send: %w", err)
	}
	return nil
}

// awaitBufferLow blocks until the DataChannel's buffered amount is
// below BufferedAmountHigh or ctx is done.
func (c *Channel) awaitBufferLow(ctx context.Context) error {
	for {
		if c.dc.BufferedAmount() < BufferedAmountHigh {
			return nil
		}
		// Arm the low watermark callback. pion only fires the callback
		// once per crossing; we re-arm each iteration.
		c.bufferedLowMu.Lock()
		if !c.bufferedLowActive {
			c.bufferedLowActive = true
			c.bufferedLowMu.Unlock()
		} else {
			c.bufferedLowMu.Unlock()
		}

		select {
		case <-c.bufferedLow:
			// loop and re-check
		case <-c.closed:
			return fmt.Errorf("transport: channel closed while waiting: %w", c.closeErr)
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

// Recv returns the next decoded frame payload (with the length prefix
// stripped) off the DataChannel. It blocks until a message arrives,
// the channel closes, or ctx is canceled.
func (c *Channel) Recv(ctx context.Context) ([]byte, error) {
	select {
	case b := <-c.recv:
		return stripLengthPrefix(b), nil
	case <-c.closed:
		return nil, fmt.Errorf("transport: channel closed: %w", c.closeErr)
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

// stripLengthPrefix removes the 4-byte big-endian length prefix from
// an inbound DataChannel message. If the message is too short or the
// prefix claims a different size, the raw bytes are returned as-is
// (backward compatibility with raw-proto senders).
func stripLengthPrefix(b []byte) []byte {
	if len(b) < 4 {
		return b
	}
	length := int(b[0])<<24 | int(b[1])<<16 | int(b[2])<<8 | int(b[3])
	if length <= 0 || 4+length > len(b) {
		// Not a valid length-prefixed message; return raw.
		return b
	}
	return b[4 : 4+length]
}

// Closed returns a channel that is closed when the underlying
// DataChannel closes.
func (c *Channel) Closed() <-chan struct{} { return c.closed }

// Close shuts the DataChannel down with a graceful close code.
// Safe to call multiple times.
func (c *Channel) Close() error {
	// closeOnce serializes the dc.Close call so the underlying SCTP
	// teardown happens exactly once. We do NOT call shutdown from
	// here: shutdown is fired by the DataChannel's OnClose callback
	// registered in New(). Calling shutdown inline used to self-
	// deadlock because shutdown also went through closeOnce.Do.
	c.closeOnce.Do(func() {
		_ = c.dc.Close()
	})
	// Defensive fallback: if for any reason OnClose does not fire
	// (e.g. the DataChannel was never fully open), ensure the
	// shutdown side-effects still run. shutdown is itself
	// idempotent under closeMu.
	c.shutdown(errors.New("channel closed by Close"))
	return nil
}

func (c *Channel) shutdown(reason error) {
	c.closeMu.Lock()
	if c.shutdownDone {
		c.closeMu.Unlock()
		return
	}
	c.shutdownDone = true
	c.closeErr = reason
	c.closeMu.Unlock()
	close(c.closed)
}

// Reassemble collects a Chunk frame into a logical message buffer. When
// all bytes are present (offset 0..total-1 covered) the assembled
// payload is returned and the per-sequence state is dropped.
//
// Reassemble is safe for concurrent use across different sequences but
// only one frame per sequence should be in flight at a time.
func (c *Channel) Reassemble(seq int32, total int, offset int, data []byte) ([]byte, bool) {
	return c.reasm.Reassemble(seq, total, offset, data)
}

// Reassembler collects Chunk frames into complete payloads. It is the
// per-connection (or per-peer) chunk assembler that the rpc.Server and
// rpc.Client both use. Splitting it out from Channel lets the rpc
// layer instantiate one without a live DataChannel (useful in tests).
type Reassembler struct {
	mu     sync.Mutex
	chunks map[int32]*assembler
}

// NewReassembler constructs an empty Reassembler.
func NewReassembler() *Reassembler {
	return &Reassembler{chunks: make(map[int32]*assembler)}
}

// Reassemble folds one chunk into the per-sequence buffer and returns
// the assembled payload plus true when complete.
func (r *Reassembler) Reassemble(seq int32, total int, offset int, data []byte) ([]byte, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	a, ok := r.chunks[seq]
	if !ok || total != a.total {
		a = &assembler{buf: make([]byte, total), total: total}
		r.chunks[seq] = a
	}
	a.lastSeen = time.Now()
	end := offset + len(data)
	if end > len(a.buf) {
		return nil, false
	}
	copy(a.buf[offset:end], data)
	a.got += len(data)
	if a.got >= a.total {
		out := a.buf
		delete(r.chunks, seq)
		return out, true
	}
	return nil, false
}

File: peerrpc-rs/transport/src/channel.rs
//! DataChannel wrapper with length-prefixed frame I/O, backpressure,
//! and chunk reassembly.
//!
//! Layering (top to bottom):
//!
//!   rpc::Client / rpc::Server
//!       │  Frame / ResponseFrame (proto messages)
//!       ▼
//!   transport::Channel::send_frame / recv
//!       │  transparent Chunk split / reassembly
//!       ▼
//!   webrtc::data_channel::RTCDataChannel (ordered, reliable)

use std::sync::Arc;

use bytes::Bytes;
use futures_util::FutureExt;
use peerrpc_protocol::{
    try_decode_frame, try_decode_response_frame, DecodeError, Frame, ResponseFrame,
    BUFFERED_AMOUNT_HIGH, MAX_FRAME_SIZE,
};
use prost::Message;
use tokio::sync::{mpsc, Mutex, Notify};
use webrtc::data_channel::RTCDataChannel;

use crate::Reassembler;

/// DataChannel label carrying the protocol version on the wire.
pub const DATACHANNEL_LABEL: &str = "peerrpc-v1";

/// Errors returned by Channel methods.
#[derive(Debug, thiserror::Error)]
pub enum ChannelError {
    #[error("transport: channel closed: {0}")]
    Closed(String),
    #[error("transport: send: {0}")]
    Send(String),
    #[error("transport: receive: {0}")]
    Recv(String),
    #[error("transport: marshal: {0}")]
    Marshal(#[from] prost::EncodeError),
    #[error("transport: decode: {0}")]
    Decode(#[from] DecodeError),
}

/// Channel is the transport-level duplex pipe backed by a WebRTC
/// DataChannel.
///
/// A Channel is safe to share across tasks for sending. It is intended
/// for a single reader (the multiplexer) but that reader may dispatch
/// decoded frames to many streams.
pub struct Channel {
    dc: Arc<RTCDataChannel>,

    /// Outbound backpressure: notified when buffered amount drops below
    /// the threshold.
    buffered_low: Arc<Notify>,

    /// Inbound queue of raw DataChannel messages.
    recv_rx: Mutex<mpsc::Receiver<Bytes>>,
    _recv_tx: mpsc::Sender<Bytes>,

    /// Chunk reassembly for large payloads split across Data frames.
    reasm: Mutex<Reassembler>,

    /// Closed signal.
    closed: Arc<Notify>,
    close_reason: Mutex<Option<String>>,
}

impl Channel {
    /// Wrap an established DataChannel.
    ///
    /// The DataChannel MUST be created with ordered=true, reliable
    /// delivery. This constructor registers OnMessage / OnClose /
    /// OnBufferedAmountLow handlers.
    pub fn new(dc: Arc<RTCDataChannel>) -> Arc<Self> {
        let (recv_tx, recv_rx) = mpsc::channel(256);
        let buffered_low = Arc::new(Notify::new());
        let closed = Arc::new(Notify::new());

        // Wire up on_message: push inbound bytes into the mpsc channel.
        let tx = recv_tx.clone();
        dc.on_message(Box::new(move |msg| {
            let tx = tx.clone();
            Box::pin(async move {
                let _ = tx.send(Bytes::copy_from_slice(&msg.data)).await;
            })
        }));

        // Wire up on_buffered_amount_low: wake any blocked senders.
        // This method is async; we spawn a task to register the handler.
        let bl = buffered_low.clone();
        let dc_clone = dc.clone();
        tokio::spawn(async move {
            dc_clone
                .on_buffered_amount_low(Box::new(move || {
                    let bl = bl.clone();
                    Box::pin(async move {
                        bl.notify_waiters();
                    })
                }))
                .await;
        });

        // Wire up on_close: signal the closed Notify.
        let cl = closed.clone();
        dc.on_close(Box::new(move || {
            let cl = cl.clone();
            Box::pin(async move {
                cl.notify_waiters();
            })
        }));

        // Set the low watermark threshold asynchronously.
        let dc_clone = dc.clone();
        tokio::spawn(async move {
            dc_clone
                .set_buffered_amount_low_threshold(BUFFERED_AMOUNT_HIGH as usize)
                .await;
        });

        Arc::new(Self {
            dc,
            buffered_low,
            recv_rx: Mutex::new(recv_rx),
            _recv_tx: recv_tx,
            reasm: Mutex::new(Reassembler::new()),
            closed,
            close_reason: Mutex::new(None),
        })
    }

    /// Send a protobuf Frame with length prefix.
    pub async fn send_frame(&self, frame: &Frame) -> Result<(), ChannelError> {
        let mut buf = Vec::new();
        Message::encode(frame, &mut buf)?;
        self.send_raw(&buf).await
    }

    /// Send a protobuf ResponseFrame with length prefix.
    pub async fn send_response_frame(&self, frame: &ResponseFrame) -> Result<(), ChannelError> {
        let mut buf = Vec::new();
        Message::encode(frame, &mut buf)?;
        self.send_raw(&buf).await
    }

    /// Send pre-encoded (length-prefixed) bytes verbatim through the
    /// DataChannel. Used by the relay, which forwards frames without
    /// inspecting them.
    ///
    /// `payload` MUST already include the 4-byte length prefix.
    pub async fn send_raw(&self, payload: &[u8]) -> Result<(), ChannelError> {
        // Check closed first (non-blocking poll).
        if self.closed.notified().now_or_never().is_some() {
            let reason = self.close_reason.lock().await.clone();
            return Err(ChannelError::Closed(
                reason.unwrap_or_else(|| "unknown".into()),
            ));
        }

        // Apply backpressure.
        self.await_buffer_low().await?;

        let bytes = Bytes::copy_from_slice(payload);
        self.dc
            .send(&bytes)
            .await
            .map_err(|e| ChannelError::Send(e.to_string()))?;

        Ok(())
    }

    /// Await until the DataChannel's buffered amount is below
    /// BUFFERED_AMOUNT_HIGH (or the channel closes).
    async fn await_buffer_low(&self) -> Result<(), ChannelError> {
        loop {
            let ba = self.dc.buffered_amount().await;
            if ba < BUFFERED_AMOUNT_HIGH as usize {
                return Ok(());
            }

            // Wait for the low watermark notification.
            tokio::select! {
                _ = self.buffered_low.notified() => {
                    // Re-check in the next iteration.
                }
                _ = self.closed.notified() => {
                    let reason = self.close_reason.lock().await.clone();
                    return Err(ChannelError::Closed(
                        reason.unwrap_or_else(|| "closed while waiting for backpressure".into()),
                    ));
                }
            }
        }
    }

    /// Receive the next raw payload (length prefix stripped) from the
    /// DataChannel. Blocks until a message arrives or the channel
    /// closes.
    ///
    /// This is the low-level receive. Higher layers typically decode
    /// the payload into a Frame or ResponseFrame.
    pub async fn recv_raw(&self) -> Result<Bytes, ChannelError> {
        let mut rx = self.recv_rx.lock().await;
        tokio::select! {
            msg = rx.recv() => {
                match msg {
                    Some(b) => Ok(Self::strip_length_prefix(&b)),
                    None => {
                        let reason = self.close_reason.lock().await.clone();
                        Err(ChannelError::Closed(
                            reason.unwrap_or_else(|| "receiver dropped".into()),
                        ))
                    }
                }
            }
            _ = self.closed.notified() => {
                let reason = self.close_reason.lock().await.clone();
                Err(ChannelError::Closed(
                    reason.unwrap_or_else(|| "channel closed".into()),
                ))
            }
        }
    }

    /// Receive and decode the next Frame from the DataChannel.
    pub async fn recv_frame(&self) -> Result<Frame, ChannelError> {
        let raw = self.recv_raw().await?;
        match try_decode_frame(&raw)? {
            Some((frame, _consumed)) => Ok(frame),
            None => Err(ChannelError::Decode(DecodeError::Oversized {
                length: raw.len(),
                max: MAX_FRAME_SIZE,
            })),
        }
    }

    /// Receive and decode the next ResponseFrame from the DataChannel.
    pub async fn recv_response_frame(&self) -> Result<ResponseFrame, ChannelError> {
        let raw = self.recv_raw().await?;
        match try_decode_response_frame(&raw)? {
            Some((frame, _consumed)) => Ok(frame),
            None => Err(ChannelError::Decode(DecodeError::Oversized {
                length: raw.len(),
                max: MAX_FRAME_SIZE,
            })),
        }
    }

    /// Fold one chunk into the per-sequence buffer and return the
    /// assembled payload when complete.
    pub async fn reassemble(
        &self,
        seq: i32,
        total: usize,
        offset: usize,
        data: &[u8],
    ) -> Option<Vec<u8>> {
        self.reasm.lock().await.reassemble(seq, total, offset, data)
    }

    /// Returns a reference to the closed Notify.
    pub fn closed_notify(&self) -> &Arc<Notify> {
        &self.closed
    }

    /// Close the DataChannel.
    pub async fn close(&self) -> Result<(), ChannelError> {
        self.dc
            .close()
            .await
            .map_err(|e| ChannelError::Send(e.to_string()))?;
        self.shutdown("closed by Close()".into()).await;
        Ok(())
    }

    async fn shutdown(&self, reason: String) {
        *self.close_reason.lock().await = Some(reason);
        self.closed.notify_waiters();
    }

    /// Strip the 4-byte big-endian length prefix from an inbound
    /// DataChannel message. If invalid, returns the raw bytes.
    fn strip_length_prefix(b: &[u8]) -> Bytes {
        if b.len() < 4 {
            return Bytes::copy_from_slice(b);
        }
        let length = u32::from_be_bytes([b[0], b[1], b[2], b[3]]) as usize;
        if length == 0 || 4 + length > b.len() {
            return Bytes::copy_from_slice(b);
        }
        Bytes::copy_from_slice(&b[4..4 + length])
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_strip_length_prefix() {
        let payload = b"hello";
        let mut prefixed = Vec::with_capacity(4 + payload.len());
        prefixed.extend_from_slice(&(payload.len() as u32).to_be_bytes());
        prefixed.extend_from_slice(payload);

        let result = Channel::strip_length_prefix(&prefixed);
        assert_eq!(&result[..], payload);
    }

    #[test]
    fn test_strip_length_prefix_too_short() {
        let result = Channel::strip_length_prefix(b"ab");
        assert_eq!(&result[..], b"ab");
    }

    #[test]
    fn test_strip_length_prefix_invalid_length() {
        let buf = vec![0xFF, 0xFF, 0xFF, 0xFF, 0x01, 0x02];
        let result = Channel::strip_length_prefix(&buf);
        assert_eq!(&result[..], &buf[..]);
    }

    #[test]
    fn test_strip_length_prefix_zero_length() {
        let buf = vec![0x00, 0x00, 0x00, 0x00];
        let result = Channel::strip_length_prefix(&buf);
        assert_eq!(&result[..], &buf[..]);
    }
}
